packages/notifications/push/amplify_push_notifications/android/src/main/kotlin/com/amazonaws/amplify/amplify_push_notifications/AmplifyPushNotificationsPlugin.kt

```diff
@@ -3,110 +3,250 @@
 import io.flutter.embedding.engine.plugins.activity.ActivityAware
 import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
+import io.flutter.plugin.common.BinaryMessenger
 import io.flutter.plugin.common.MethodCall
+import io.flutter.plugin.common.MethodChannel
 import io.flutter.plugin.common.MethodChannel.MethodCallHandler
 import io.flutter.plugin.common.MethodChannel.Result
 import io.flutter.plugin.common.PluginRegistry
+import kotlinx.coroutines.*
+
+
+private const val TAG = "AmplifyPushNotificationsPlugin"
 
 /** AmplifyPushNotificationsPlugin */
-class AmplifyPushNotificationsPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
+open class AmplifyPushNotificationsPlugin : FlutterPlugin, MethodCallHandler, ActivityAware,
     PluginRegistry.NewIntentListener {
+    private companion object {
+        const val TAG = "AmplifyPushNotificationsPlugin"
+
+        /**
+         * The scope in which to spawn tasks which should not be awaited from the main thread.
+         */
+        val scope =
+            CoroutineScope(Dispatchers.IO) + CoroutineName("amplify_flutter.PushNotifications")
+    }
 
+    /**
+     * The user's main activity.
+     */
+    private var mainActivity: Activity? = null
+
+    /**
+     * The plugin binding for [mainActivity], used to manage lifecycle callback registration.
+     */
     private var activityBinding: ActivityPluginBinding? = null
-    companion object {
-        private val TAG = "AmplifyPushNotificationsPlugin"
+
+    /**
+     * The application context.
+     */
+    private var applicationContext: Context? = null
+
+    /**
+     * Shared Preference used to persist callback handles.
+     */
+    private lateinit var sharedPreferences: SharedPreferences
+    private lateinit var channel: MethodChannel
+
+    /**
+     * Main engine's binary messenger holder.
+     */
+    private var mainBinaryMessenger: BinaryMessenger? = null
+
+    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
+        mainBinaryMessenger = flutterPluginBinding.binaryMessenger
+        applicationContext = flutterPluginBinding.applicationContext
+        sharedPreferences =
+            applicationContext!!.getSharedPreferences(
+                PushNotificationConstants.SHARED_PREFERENCES_KEY,
+                Context.MODE_PRIVATE
+            )
+        channel = MethodChannel(
+            flutterPluginBinding.binaryMessenger, PushNotificationConstants.METHOD_CHANNEL
+        )
+        channel.setMethodCallHandler(this)
+    }
+
+    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
+        channel.setMethodCallHandler(null)
     }
 
     override fun onAttachedToActivity(binding: ActivityPluginBinding) {
+        mainActivity = binding.activity
         activityBinding = binding
-        onNewIntent(binding.activity.intent)
+        mainBinaryMessenger?.let {
```

Comment:
Since the binary messenger is only available in `onAttachedToEngine`, these should probably only be initialized there

Reply:
Same for deinit - move to `onDetachedFromEngine`

Reply:
Yes, that was the initial approach but when we create the Background Isolate upon receiving a notification in the background/killed state, `onAttachedToEngine` gets called again with the Background Isolate's binaryMessenger. We don't want to initialize the stream handlers with that messenger but rather keep it attached to the main Flutter Engine that displays the app. Which is why I store the messenger and only initialize when the plugin is attached to the activity.

Reply:
Both the [situations](https://api.flutter.dev/javadoc/io/flutter/embedding/engine/plugins/activity/ActivityAware.html#:~:text=1%20of%202-,situations,-%3A) mentioned in the API doc for `onAttachedToActivity` indicates that this order of connection `Plugin -> FlutterEngine -> Activity` 
So, no it can't be called before a FlutterEngine is present.